### DISCUSSION TITLE
Any way to create custom subscriber?

### DISCUSSION THREAD

**Discussion by BigBadAlien** [klis87/redux-saga-requests#496] [Q&A] [answered]
We have a pretty detailed description of [how to create custom driver](https://github.com/klis87/redux-requests/blob/b9921b6e02cee56617690e108556e8450b5e1a47/docusaurus/docs/drivers/using-drivers.md#writing-your-own-driver) but does exist any way to create such a custom driver for event-based transport?

The current GraphQL subscriber looks like

```
export const createSubscriber: (config: {
  url: string;
  WS?: any;
  lazy?: boolean;
  heartbeatTimeout?: number;
  reconnectTimeout?: number;
  useHeartbeat?: boolean;
}) => Subscriber;
```

It means that the subscriber is WebSocket based and a URL is required. But what if we want to use e.g. EventEmitter as a source events, do we have any way for it?

**Comment by klis87** [comment, marked as answer]
Interesting topic. Indeed a custom driver is about implementing not only some custom behaviour, but you can do anything you like, even connect to an external API library, or work on pure promises, whatever you want. Subscribers are a little different, like you noticed, it already assumes WebSocket protocol. Creating subscriber of your own is about making it compatible with your WS server, be it GraphQL or anything else. It would be cool to make it support anything, but currently only WS are possible.

But, there is a hope. Perhaps it would be possible to create a WS adapter class? As you can see, you can pass `WS` to `createSubscriber`. What about trying to make a class implementing WS interface, but under the hood using something else, like `EventEmitter`?

**Reply by BigBadAlien** [reply]
>  create a WS adapter class
I'll try it, thanks
